kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/MergeQueries.java

```diff
@@ -124,211 +129,312 @@ private void mergeFlush(
+        VALUES (
+          src.key,
+          [CAST(CAST(DATE(src.partitionTime) AS DATE) AS TIMESTAMP),]
+          src.value.<field>[, src.value.<field>...]
+        );
+   */
+  private String upsertMergeFlushQuery(
+      TableId intermediateTable, TableId destinationTable, int batchNumber, Schema intermediateSchema
+  ) {
+    List<String> keyFields = listFields(
+        intermediateSchema.getFields().get(INTERMEDIATE_TABLE_KEY_FIELD_NAME).getSubFields(),
+        INTERMEDIATE_TABLE_KEY_FIELD_NAME + "."
+    );
 
-    upsert only:
+    List<String> valueColumns = valueColumns(intermediateSchema);
+
+    final String key = INTERMEDIATE_TABLE_KEY_FIELD_NAME;
+    final String i = INTERMEDIATE_TABLE_ITERATION_FIELD_NAME;
+    final String value = INTERMEDIATE_TABLE_VALUE_FIELD_NAME;
+    final String batch = INTERMEDIATE_TABLE_BATCH_NUMBER_FIELD;
+
+    return "MERGE " + table(destinationTable) + " "
+        + "USING ("
+          + "SELECT * FROM ("
+            + "SELECT ARRAY_AGG("
+              + "x ORDER BY " + i + " DESC LIMIT 1"
+            + ")[OFFSET(0)] src "
+            + "FROM " + table(intermediateTable) + " x "
+            + "WHERE " + batch + "=" + batchNumber + " "
+            + "GROUP BY " + String.join(", ", keyFields)
+          + ")"
+        + ") "
+        + "ON `" + destinationTable.getTable() + "`." + keyFieldName + "=src." + key + " "
+        + "WHEN MATCHED "
+          + "THEN UPDATE SET " + valueColumns.stream().map(col -> col + "=src." + value + "." + col).collect(Collectors.joining(", ")) + " "
+        + "WHEN NOT MATCHED "
+          + "THEN INSERT ("
+            + keyFieldName + ", "
+            + partitionTimePseudoColumn()
+            + String.join(", ", valueColumns) + ") "
+          + "VALUES ("
+            + "src." + key + ", "
+            + partitionTimeValue()
+            + valueColumns.stream().map(col -> "src." + value + "." + col).collect(Collectors.joining(", "))
+          + ");";
+  }
 
-    MERGE `<dataset>`.`<destinationTable>`
-    USING (
-      SELECT * FROM (
-        SELECT ARRAY_AGG(
-          x ORDER BY partitionTime DESC LIMIT 1
-        )[OFFSET(0)] src
-        FROM `<dataset>`.`<intermediateTable>` x
-        WHERE batchNumber=<batchNumber>
-        GROUP BY key.<field>[, key.<field>...]
-      )
-    )
-    ON `<destinationTable>`.<keyField>=`src`.key
-    WHEN MATCHED
-      THEN UPDATE SET <valueField>=`src`.value.<field>[, <valueField=`src.value.<field>...]
-    WHEN NOT MATCHED
-      THEN INSERT (<keyField, _PARTITIONTIME, <valueField[, <valueField])
+  /*
+        Delete-only is the trickiest mode. Naively, we could just run a MERGE using the intermediate
+      table as a source and sort in ascending order of iteration. However, this would miss an edge
+      case where, for a given key, a non-tombstone record is sent and then followed by a tombstone,
+      and would result in all rows with that key being deleted from the table, followed by an
```

Comment:
Could we retain the simpler query but do something in code to make this edge case work? Maybe explicitly trigger a flush when seeing a tombstone or something? Just thinking out load. We would potentially trade off some performance due to the increased number of flushes but because of the simpler query. may gain back some from the query engine. There might also be some consequences to consider due to the eventual consistent nature of BigQuery?

Reply:
Yeah, we can definitely consider this as a potential optimization for the connector down the road. I'd like to keep things uniform right now if possible (which is why, for example, we use the exact same intermediate table schema for the different combinations of upsert/delete modes) and not introduce more gotchas as far as merge query frequency goes. If this query works as expected, we may not have to alter anything and it can gather dust quietly without any further tweaks.